addons/advanced_ballistics/functions/fnc_handleFired.sqf

```diff
@@ -21,22 +21,45 @@ TRACE_10("firedEH:",_unit, _weapon, _muzzle, _mode, _ammo, _magazine, _projectil
 
 if (!(_ammo isKindOf "BulletBase")) exitWith {};
 if (!alive _projectile) exitWith {};
-if (_unit distance ACE_player > GVAR(simulationRadius)) exitWith {};
 if (underwater _unit) exitWith {};
 
-private _abort = GVAR(disabledInFullAutoMode) && {getNumber(configFile >> "CfgWeapons" >> _weapon >> _mode >> "autoFire") == 1};
-if (!local _unit && {!_abort && !GVAR(simulateForEveryone)}) then {
-    // The shooter is non local -> abort unless we match an exception rule        
-    _abort = !GVAR(simulateForGroupMembers) || {group ACE_player != group _unit};
-    if (_abort && GVAR(simulateForSnipers)) then {
-        if (currentWeapon _unit == primaryWeapon _unit && count primaryWeaponItems _unit > 2) then {
-            private _opticsName = (primaryWeaponItems _unit) select 2;
+private _abort = !local _unit;
+if (_abort) then {
+    private _bulletVelocity = velocity _projectile;
+    private _muzzleVelocity = vectorMagnitude _bulletVelocity;
+    
+    private _maxRange = uiNamespace getVariable format[QGVAR(maxRange_%1), _ammo];
+    if (isNil "_maxRange") then {
+        private _airFriction = getNumber(configFile >> "CfgAmmo" >> _ammo >> "airFriction");
+        private _maxTime = ((_muzzleVelocity - BULLET_TRACE_MIN_VELOCITY) / (BULLET_TRACE_MIN_VELOCITY * -_airFriction * _muzzleVelocity)) max getNumber(configFile >> "CfgAmmo" >> _ammo >> "tracerEndTime");
+        private _maxRange = if (_airFriction < 0) then {
+            -ln(1 - _airFriction * _muzzleVelocity * _maxTime) / _airFriction
+        } else {
+            _muzzleVelocity * _maxTime
+        };
+        uiNamespace setVariable [format[QGVAR(maxRange_%1), _ammo], _maxRange];
+    };
+    if (ACE_player distance _unit > _maxRange && {ACE_player distance ((getPosASL _unit) vectorAdd ((vectorNormalized _bulletVelocity) vectorMultiply _maxRange)) > _maxRange}) exitWith {};
+    
+    private _ammoCount = _unit ammo _muzzle;
```

Comment:
I think we want to add +1 here as first shot is 29 and last is 0 (the ammo is already removed by the time firedEH runs)

Reply:
Right!